dftimewolf/lib/processors/gcp_cloud_resource_tree.py

```diff
@@ -0,0 +1,1002 @@
+          continue
+
+        # Check if a resource with the same ID already exist in the
+        # self._resources_dict dictionary
+        resource = self.resources_dict.get(response.get('targetId'))
+
+        if not resource:
+          resource = Resource()
+          resource.id = response.get('targetId')
+
+        # compute.disks.createSnapshot is a special case where the
+        # "resourceName" is just the name and not the full name with the
+        # project, zone, type and name
+        if log_message_type.startswith('compute.disks.createSnapshot'):
+          resource = Resource()
+          source_disk_id = response.get('targetId')
+          resource.name = request.get('name')
+          # GCP Log issue, the resource.type is set to 'gce_disk' so i am
+          # setting it manually to gce_snapshot
+          resource.type = 'gce_snapshot'
+          resource.project_id = log_message.get('resource',
+                                                {}).get('labels',
+                                                        {}).get('project_id')
+          resource.zone = 'global'
+
+          # For creation of a snapshot, its a special case where the targetId is
+          # of the source disk and not the snapshot being created. And the
+          # targetLink points to the source disk. The id of the snapshot
+          # is not present in the log message.
+          # We need to unset the id which was set earlier to targetId
+          matched_resource = self._FindResource(resource.name, resource.type)
+          if matched_resource:
+            resource = matched_resource
+          else:
+            resource.id = source_disk_id + "-snapshot"
+
+        else:
+          resource.resource_name = resource.resource_name or proto_payload.get(
+              'resourceName', '')
+
+        # In case the message is an insert message
+        if log_message_type.endswith('insert') or log_message_type.endswith(
+            'createSnapshot'):
+          if request_metadata:
+            resource.creator_ip_address = request_metadata.get('callerIp')
+            resource.creator_useragent = request_metadata.get(
+                'callerSuppliedUserAgent')
+          self._ParseInsertLogMessage(resource, request, response)
+
+        elif log_message_type.endswith('delete'):
+          resource.deletion_timestamp = response.get('insertTime')
+          resource.deleted_by = response.get('user')
+          if request_metadata:
+            resource.deleter_ip_address = request_metadata.get('callerIp')
+            resource.deleter_useragent = request_metadata.get(
+                'callerSuppliedUserAgent')
+
+      else:
+        self.PublishMessage(f'Type {log_message_type} not supported')
+        resource = None
+
+      if resource:
+        if resource.id:
+          self.resources_dict[resource.id] = resource
+
+  def _ParseInsertLogMessage(self, resource: Resource, request: Dict[str, Any],
```

Comment:
please add some description that this function changes the passed Resource object in-place with the information parsed from request and response arguments

Reply:
Done